torch/_inductor/bounds.py

```diff
@@ -22,12 +22,22 @@ class BoundVars:
 
     def __init__(self, loop_body: LoopBody):
         self.loop_body = loop_body
-        self.replacement_vals = {
-            k: ValueRanges(0, v - 1)
-            if not free_symbols(v)
-            else ValueRanges(2, math.inf)
-            for k, v in loop_body.var_ranges.items()
-        }
+        # Variable substitutions for any constrained ranges:
+        subs = []
+        tracing_context = torch._guards.TracingContext.get()
+        if tracing_context:
+            subs = [
+                (k, v.upper)
```

Comment:
we should have separate substitutions for upper and lower bound

Reply:
@eellison - thanks; meant to ask you about the lower bound handling. I might not understand exactly what we're doing here. If we have a range that's a symbolic expression of `4 * s0`, for example, but we also have a constrained `var_range` available like `s0: ValueRanges(lower=2, upper=100)` ... I guess I assumed that only the upper bound was relevant for the int64 -> int32 optimization and we could use `4 * 100 = 400` for the upper bound. How is the lower bound relevant exactly? Do you just mean that we can also further constrain the lower bound on this line (and not assume 0)?

`   ValueRanges(0, s - 1)`

Reply:
If `x` has bounds `[a, b]`, then `-x` has bounds `[-b, -a]`, so you need to track both.